//
// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto2";

package testing;

option java_package = "com.google.privacy.differentialprivacy.proto.testing";

// The following protos are intended to store the parameters required to asses
// the DP library based on the statistical tests.

enum NoiseType {
  UNDEFINED_NOISE = 0;
  LAPLACE = 1;
  GAUSSIAN = 2;
}

enum AggregationType {
  UNDEFINED_AGGREGATION = 0;
  COUNT = 1;
  SUM = 2;
  MEAN = 3;
  STDV = 5;
  VARIANCE = 6;
  QUANTILE = 7;
  SELECT_PARTITIONS = 9;
}

message DataRecord {
  optional int32 user = 1;
  optional int32 partition = 2;
  optional double double_value = 3;
}

// Parameters required for running a closeness test
message ClosenessTestParameters {
  // mean of the distribution that is used as a reference
  optional double mean = 1;
  // variance of the distribution that is used as a reference
  optional double variance = 2;
  // l2 distance deemed acceptable between the distribution that is being tested
  // for and the reference distribution
  optional double l2_tolerance = 3;
  // parameter quantifying the amount of discretization that is applied to
  // samples before testing
  optional double granularity = 4;
}

// Parameters required for running an approximate DP test
message DpTestParameters {
  // parameter defining the epsilon the statistical DP test tries to verify
  // (this epsilon should not be used for drawing samples; the epsilon for
  // drawing samples is defined in the respective sampling proto messages)
  optional double epsilon = 1;
  // parameter defining the delta the statistical DP test tries to verify (this
  // delta should not be used for drawing samples; the delta for drawing samples
  // is defined in the respective sampling proto messages)
  optional double delta = 2;
  // parameter specifying the acceptable deviation from delta
  optional double delta_tolerance = 3;
  // parameter quantifying the amount of discretization that is applied to
  // samples before testing
  optional double granularity = 4;
  // parameter defining the number of buckets to which samples are mapped before
  // testing
  optional int32 num_of_buckets = 5;
}

// Parameters required for drawing samples from the add noise method of the DP
// library.
message NoiseSamplingParameters {
  optional int32 number_of_samples = 1;
  optional int32 l0_sensitivity = 2;
  optional double linf_sensitivity = 3;
  optional double epsilon = 4;
  optional double delta = 5;
  optional double raw_input = 6;
}

// Parameters required for drawing samples from the count aggregation of the DP
// library.
message CountSamplingParameters {
  optional int32 number_of_samples = 1;
  optional NoiseType noise_type = 2;
  optional int32 max_partitions_contributed = 3;
  optional double epsilon = 4;
  optional double delta = 5;
  repeated int64 raw_increment_by = 6;
  repeated int64 neighbour_raw_increment_by = 7;
}

// Parameters required for drawing samples from the bounded sum aggregation of
// the DP library.
message BoundedSumSamplingParameters {
  optional int32 number_of_samples = 1;
  optional NoiseType noise_type = 2;
  optional int32 max_partitions_contributed = 3;
  optional double lower_bound = 4;
  optional double upper_bound = 5;
  optional double epsilon = 6;
  optional double delta = 7;
  repeated double raw_entry = 8;
  repeated double neighbour_raw_entry = 9;
}

// Parameters required for drawing samples from the bounded mean aggregation of
// the DP library.
message BoundedMeanSamplingParameters {
  optional int32 number_of_samples = 1;
  optional NoiseType noise_type = 2;
  optional int32 max_contributions_per_partition = 3;
  optional int32 max_partitions_contributed = 4;
  optional double lower_bound = 5;
  optional double upper_bound = 6;
  optional double epsilon = 7;
  optional double delta = 8;
  repeated double raw_entry = 9;
  repeated double neighbour_raw_entry = 10;
}

// Parameters required for drawing samples from the bounded stdv and bounded
// variance aggregation of the DP library.
message BoundedStdvSamplingParameters {
  optional int32 number_of_samples = 1;
  optional NoiseType noise_type = 2;
  optional int32 max_contributions_per_partition = 3;
  optional int32 max_partitions_contributed = 4;
  optional double lower_bound = 5;
  optional double upper_bound = 6;
  optional double epsilon = 7;
  optional double delta = 8;
  repeated double raw_entry = 9;
  repeated double neighbour_raw_entry = 10;
}

// Parameters required for drawing samples from the bounded quantiles
// aggregation of the DP library.
message BoundedQuantilesSamplingParameters {
  optional int32 number_of_samples = 1;
  optional NoiseType noise_type = 2;
  repeated double rank = 3;
  optional int32 max_contributions_per_partition = 4;
  optional int32 max_partitions_contributed = 5;
  optional double lower_bound = 6;
  optional double upper_bound = 7;
  optional int32 tree_height = 8;
  optional int32 branching_factor = 9;
  optional double epsilon = 10;
  optional double delta = 11;
  repeated double raw_entry = 12;
  repeated double neighbour_raw_entry = 13;
}

// Parameters required for drawing sample bounds from the approximate bounds
// calculator of the DP library.
message ApproximateBoundsSamplingParameters {
  optional int32 number_of_samples = 1;
  optional NoiseType noise_type = 2;
  optional int32 max_contributions = 3;
  optional int32 number_of_bins = 4;
  optional double scale = 5;
  optional double base = 6;
  optional double epsilon = 7;
  optional double delta = 8;
  repeated double raw_entry = 9;
  repeated double neighbour_raw_entry = 10;
}

// General parameters for drawing samples, intended for testing various
// functionality of higher level tools.
message GeneralSamplingParameters {
  optional int32 number_of_samples = 1;
  repeated AggregationType aggregation_type = 2;
  optional int32 max_partitions_contributed = 3;
  optional int32 max_contributions_per_partition = 4;
  optional double lower_bound = 5;
  optional double upper_bound = 6;
  optional double epsilon = 7;
  optional double delta = 8;
  repeated DataRecord record = 9;
  repeated DataRecord neighbour_record = 10;
}

// Parameters specifying a voting mechanism for evaluating multiple runs of a
// particular statistical test. This is intended to reduce the probability of
// error compared to evaluating just a single run of the respective test.
message VotingParameters {
  optional int32 number_of_votes = 1;
}

// Parameters quantifying notions of specificity that a particular closeness
// test should satisfy.
message ClosenessTestValidityParameters {
  // The minimum difference between the mean of two sets of samples relative to
  // their standard deviation that should be rejected.
  optional double shift_specificity = 1;
  // The minimum difference in scale between the variance of two sets of samples
  // that should be rejected.
  optional double scale_specificity = 2;
}

// Parameters quantifying notions of specificity that a particular approximate
// DP test should satisfy.
message DpTestValidityParameters {
  // The minimum distance between two sets of samples relative to their
  // anticipated distance that should be rejected.
  optional double distance_specificity = 1;
  // The minimum rate of catastrophic privacy failures relative to the delta
  // tolerance of the DP test that should be rejected.
  optional double failure_specificity = 2;
  // The minimum ratio between an epsilon that should be accepted and an epsilon
  // that should be rejected, i.e., if epsilon is the smallest test parameter
  // for which we expect the test to accept, a test paramter of
  // epsilon / epsilon_specificity should be rejected.
  optional double epsilon_specificity = 3;
  // The minimum ratio between a delta that should be accepted and a delta that
  // should be rejected, i.e., if delta is the test privacy parameter for which
  // which we expect the test to accept, a test paramter of
  // delta / delta_specificity should be rejected.
  optional double delta_specificity = 4;
}

message DistributionClosenessTestCase {
  optional string name = 1;
  optional ClosenessTestParameters closeness_test_parameters = 2;
  optional NoiseSamplingParameters noise_sampling_parameters = 3;
}

message DistributionDpTestCase {
  optional string name = 1;
  optional DpTestParameters dp_test_parameters = 2;
  optional NoiseSamplingParameters noise_sampling_parameters = 3;
}

message CountDpTestCase {
  optional string name = 1;
  optional DpTestParameters dp_test_parameters = 2;
  optional CountSamplingParameters count_sampling_parameters = 3;
}

message BoundedSumDpTestCase {
  optional string name = 1;
  optional DpTestParameters dp_test_parameters = 2;
  optional BoundedSumSamplingParameters bounded_sum_sampling_parameters = 3;
}

message BoundedMeanDpTestCase {
  optional string name = 1;
  optional DpTestParameters dp_test_parameters = 2;
  optional BoundedMeanSamplingParameters bounded_mean_sampling_parameters = 3;
}

message BoundedStdvDpTestCase {
  optional string name = 1;
  optional DpTestParameters dp_test_parameters = 2;
  optional BoundedStdvSamplingParameters bounded_stdv_sampling_parameters = 3;
}

message BoundedQuantilesDpTestCase {
  optional string name = 1;
  optional DpTestParameters dp_test_parameters = 2;
  optional BoundedQuantilesSamplingParameters
      bounded_quantiles_sampling_parameters = 3;
}

message ApproximateBoundsDpTestCase {
  optional string name = 1;
  optional DpTestParameters dp_test_parameters = 2;
  optional ApproximateBoundsSamplingParameters
      approximate_bounds_sampling_parameters = 3;
}

message DistributionClosenessTestCaseCollection {
  repeated DistributionClosenessTestCase distribution_closeness_test_case = 1;
  optional VotingParameters voting_parameters = 2;
  optional ClosenessTestValidityParameters validity_parameters = 3;
}

message DistributionDpTestCaseCollection {
  repeated DistributionDpTestCase distribution_dp_test_case = 1;
  optional VotingParameters voting_parameters = 2;
  optional DpTestValidityParameters validity_parameters = 3;
}

message CountDpTestCaseCollection {
  repeated CountDpTestCase count_dp_test_case = 1;
  optional VotingParameters voting_parameters = 2;
  optional DpTestValidityParameters validity_parameters = 3;
}

message BoundedSumDpTestCaseCollection {
  repeated BoundedSumDpTestCase bounded_sum_dp_test_case = 1;
  optional VotingParameters voting_parameters = 2;
  optional DpTestValidityParameters validity_parameters = 3;
}

message BoundedMeanDpTestCaseCollection {
  repeated BoundedMeanDpTestCase bounded_mean_dp_test_case = 1;
  optional VotingParameters voting_parameters = 2;
  optional DpTestValidityParameters validity_parameters = 3;
}

message BoundedStdvDpTestCaseCollection {
  repeated BoundedStdvDpTestCase bounded_stdv_dp_test_case = 1;
  optional VotingParameters voting_parameters = 2;
  optional DpTestValidityParameters validity_parameters = 3;
}

message BoundedQuantilesDpTestCaseCollection {
  repeated BoundedQuantilesDpTestCase bounded_quantiles_dp_test_case = 1;
  optional VotingParameters voting_parameters = 2;
  optional DpTestValidityParameters validity_parameters = 3;
}

message ApproximateBoundsDpTestCaseCollection {
  repeated ApproximateBoundsDpTestCase approximate_bounds_dp_test_case = 1;
  optional VotingParameters voting_parameters = 2;
  optional DpTestValidityParameters validity_parameters = 3;
}
